UnityMainThreadDispatcher.cs

```diff
@@ -29,12 +29,18 @@ public class UnityMainThreadDispatcher : MonoBehaviour {
 	private static readonly Queue<Action> _executionQueue = new Queue<Action>();
 
 	public void Update() {
-		lock(_executionQueue) {
-			while (_executionQueue.Count > 0) {
-				_executionQueue.Dequeue().Invoke();
-			}
-		}
-	}
+        // Make local copy to avoid holding the lock while the actions execute.
+        Queue<Action> queueCopy;
+		lock (_executionQueue)
+		{
+			queueCopy = new Queue<Action>(_executionQueue);
+			_executionQueue.Clear();
```

Comment:
What is the reason for clearing the queue here? Because it's being moved over to an entirely new queue? I get the idea but it seems a lot can go wrong here.

Reply:
Can you say more about what could go wrong?

The copy-and-clear step is effectively the same as the previous while-not-empty-dequeue; we're just moving the Invoke step outside of the lock. The invariant from the original implementation should still be the same - that each action in _executionQueue when Update is called gets executed on the main thread and that those actions are no longer in the queue.

This just means background threads only wait for a queue to copy instead of arbitrarily expensive actions to execute. (Admittedly, if the actions are that expensive, they likely shouldn't be executing on the main thread, but that's a different issue).